Guardian Project Tor Android JNI/build adaptation, retained without runtime changes.
Upstream: https://github.com/bitmold/tor
Base: 8fe2745aee2c35116ff1081f19b2952418b43069 (Tor 0.4.9.14)
Donor: 4a522a3ca491c6c04d4a44cb006b92101c42efc1
Source: https://github.com/bitmold/tor/compare/8fe2745aee2c35116ff1081f19b2952418b43069...4a522a3ca491c6c04d4a44cb006b92101c42efc1.diff

BashKitten selection: retain the donor's Android runtime and build changes;
omit upstream CI, funding and disabled-test changes. Apply to pristine official
Tor 0.4.9.15 only in the external build directory. Original copyright notices
remain in the imported JNI sources; see upstream/LICENSE and Tor LICENSE.

diff --git a/changes/ticket31882 b/changes/ticket31882
new file mode 100644
index 00000000000..bcf17ce2a8e
--- /dev/null
+++ b/changes/ticket31882
@@ -0,0 +1,7 @@
+  o Minor feature (configure, build system):
+    - The standard __ANDROID__ macro is now used for platform detection.
+    - --enable-android provides the standard, minimal build configuration
+      for creating Tor binaries to run either as a shared library or a daemon.
+      It enables PIC, disables system torrc, systemd, and all doc generation.
+      This flag is not required to build for Android.
+      Closes ticket 31882.
diff --git a/configure.ac b/configure.ac
index 6b3cefd5292..6e200c9bef5 100644
--- a/configure.ac
+++ b/configure.ac
@@ -96,6 +96,8 @@ AC_ARG_ENABLE(nss,
    AS_HELP_STRING(--enable-nss, [Use Mozilla's NSS TLS library. (EXPERIMENTAL)]))
 AC_ARG_ENABLE(pic,
    AS_HELP_STRING(--enable-pic, [Build Tor's binaries as position-independent code, suitable to link as a library.]))
+AC_ARG_ENABLE(jni,
+   AS_HELP_STRING(--enable-jni, [enable Java JNI API feature]))
 
 AC_ARG_ENABLE(missing-doc-warnings,
    AS_HELP_STRING(--enable-missing-doc-warnings, [Tell doxygen to warn about missing documentation. Makes doxygen warnings nonfatal.]))
@@ -115,6 +117,39 @@ fi
 AC_SUBST(DOXYGEN_FATAL_WARNINGS)
 AC_SUBST(DOXYGEN_WARN_ON_MISSING)
 
+dnl Enable minimal, standard Android build
+AC_ARG_ENABLE(android,
+     AS_HELP_STRING(--enable-android, [setup build for Android exeutable/shared library]))
+
+if test "x$enable_android" = "xyes"; then
+  asciidoc=false
+  enable_html_manual=no
+  enable_manpage=no
+  enable_jni=yes
+  enable_pic=yes
+  enable_system_torrc=no
+  enable_tool_name_check=no
+  have_systemd=no
+
+  need_cross_compiling_host=no
+  if test "x$cross_compiling" != "xyes"; then
+       need_cross_compiling_host=yes
+  fi
+  case "$host" in
+       *-linux-android*) ;;
+       *)
+       need_cross_compiling_host=yes
+       ;;
+  esac
+  if test "x$need_cross_compiling_host" = "xyes"; then
+       AC_MSG_ERROR([
+
+ Android builds are cross-compiled, therefore require setting --host= which
+ can be derived from the NDK ABIs: https://developer.android.com/ndk/guides/abis
+       ])
+  fi
+fi
+
 if test "x$enable_coverage" != "xyes" -a "x$enable_asserts_in_tests" = "xno" ; then
     AC_MSG_ERROR([Can't disable assertions outside of coverage build])
 fi
@@ -170,7 +205,7 @@ AC_ARG_ENABLE(asciidoc,
         "yes") asciidoc=true ;;
         "no")  asciidoc=false ;;
         *) AC_MSG_ERROR(bad value for --disable-asciidoc) ;;
-      esac], [asciidoc=true])
+      esac], [test "x$asciidoc" = "xtrue" && asciidoc=true])
 
 # systemd notify support
 AC_ARG_ENABLE(systemd,
@@ -1868,6 +1903,23 @@ AC_CHECK_TYPES([rlim_t], , ,
 #endif
 ])
 
+if test "x$enable_jni" = "xyes"; then
+  AX_JNI_INCLUDE_DIR
+  if test -n "$JNI_INCLUDE_DIRS"; then
+    AC_DEFINE(HAVE_JNI_H,1,[Have jni.h])
+    for JNI_INCLUDE_DIR in $JNI_INCLUDE_DIRS; do
+      CPPFLAGS="$CPPFLAGS -I$JNI_INCLUDE_DIR"
+    done
+  fi
+  dnl Enable Android only standard libs
+  case "$host" in
+    *android*)
+      AC_CHECK_LIB([log], [__android_log_write], [LIBS="-llog $LIBS"])
+      ;;
+  esac
+fi
+AM_CONDITIONAL(USE_JNI, test -n "$JNI_INCLUDE_DIRS")
+
 AX_CHECK_SIGN([time_t],
        [ : ],
        [ : ], [
diff --git a/m4/ax_jni_include_dir.m4 b/m4/ax_jni_include_dir.m4
new file mode 100644
index 00000000000..ae7a5f042be
--- /dev/null
+++ b/m4/ax_jni_include_dir.m4
@@ -0,0 +1,151 @@
+# ===========================================================================
+#    https://www.gnu.org/software/autoconf-archive/ax_jni_include_dir.html
+# ===========================================================================
+#
+# SYNOPSIS
+#
+#   AX_JNI_INCLUDE_DIR
+#
+# DESCRIPTION
+#
+#   AX_JNI_INCLUDE_DIR finds include directories needed for compiling
+#   programs using the JNI interface.
+#
+#   JNI include directories are usually in the Java distribution. This is
+#   deduced from the value of $JAVA_HOME, $JAVAC, or the path to "javac", in
+#   that order. When this macro completes, a list of directories is left in
+#   the variable JNI_INCLUDE_DIRS.
+#
+#   Example usage follows:
+#
+#     AX_JNI_INCLUDE_DIR
+#
+#     for JNI_INCLUDE_DIR in $JNI_INCLUDE_DIRS
+#     do
+#             CPPFLAGS="$CPPFLAGS -I$JNI_INCLUDE_DIR"
+#     done
+#
+#   If you want to force a specific compiler:
+#
+#   - at the configure.in level, set JAVAC=yourcompiler before calling
+#   AX_JNI_INCLUDE_DIR
+#
+#   - at the configure level, setenv JAVAC
+#
+#   This macro depends on AC_CANONICAL_HOST which requires that config.guess
+#   and config.sub be distributed along with the source code.  See autoconf
+#   manual for details.
+#
+#   Note: This macro can work with the autoconf M4 macros for Java programs.
+#   This particular macro is not part of the original set of macros.
+#
+# LICENSE
+#
+#   Copyright (c) 2008 Don Anderson <dda@sleepycat.com>
+#
+#   Copying and distribution of this file, with or without modification, are
+#   permitted in any medium without royalty provided the copyright notice
+#   and this notice are preserved. This file is offered as-is, without any
+#   warranty.
+
+#serial 15
+
+AU_ALIAS([AC_JNI_INCLUDE_DIR], [AX_JNI_INCLUDE_DIR])
+AC_DEFUN([AX_JNI_INCLUDE_DIR],[
+
+AC_REQUIRE([AC_CANONICAL_HOST])
+
+JNI_INCLUDE_DIRS=""
+
+if test "x$JAVA_HOME" != x; then
+	_JTOPDIR="$JAVA_HOME"
+else
+	if test "x$JAVAC" = x; then
+		JAVAC=javac
+	fi
+	AC_PATH_PROG([_ACJNI_JAVAC], [$JAVAC], [no])
+	if test "x$_ACJNI_JAVAC" = xno; then
+		AC_MSG_ERROR([cannot find JDK; try setting \$JAVAC or \$JAVA_HOME])
+	fi
+	_ACJNI_FOLLOW_SYMLINKS("$_ACJNI_JAVAC")
+	_JTOPDIR=`echo "$_ACJNI_FOLLOWED" | sed -e 's://*:/:g' -e 's:/[[^/]]*$::'`
+fi
+
+case "$host_os" in
+        darwin*)        # Apple Java headers are inside the Xcode bundle.
+            macos_version=$(sw_vers -productVersion | sed -n -e 's/^@<:@0-9@:>@*.\(@<:@0-9@:>@*\).@<:@0-9@:>@*/\1/p')
+            if @<:@ "$macos_version" -gt "7" @:>@; then
+                _JTOPDIR="$(xcrun --show-sdk-path)/System/Library/Frameworks/JavaVM.framework"
+                _JINC="$_JTOPDIR/Headers"
+            else
+                _JTOPDIR="/System/Library/Frameworks/JavaVM.framework"
+                _JINC="$_JTOPDIR/Headers"
+            fi
+            ;;
+        *) _JINC="$_JTOPDIR/include";;
+esac
+_AS_ECHO_LOG([_JTOPDIR=$_JTOPDIR])
+_AS_ECHO_LOG([_JINC=$_JINC])
+
+# On Mac OS X 10.6.4, jni.h is a symlink:
+# /System/Library/Frameworks/JavaVM.framework/Versions/Current/Headers/jni.h
+# -> ../../CurrentJDK/Headers/jni.h.
+AC_CACHE_CHECK(jni headers, ac_cv_jni_header_path,
+[
+  if test -f "$_JINC/jni.h"; then
+    ac_cv_jni_header_path="$_JINC"
+    JNI_INCLUDE_DIRS="$JNI_INCLUDE_DIRS $ac_cv_jni_header_path"
+  else
+    _JTOPDIR=`echo "$_JTOPDIR" | sed -e 's:/[[^/]]*$::'`
+    if test -f "$_JTOPDIR/include/jni.h"; then
+      ac_cv_jni_header_path="$_JTOPDIR/include"
+      JNI_INCLUDE_DIRS="$JNI_INCLUDE_DIRS $ac_cv_jni_header_path"
+    else
+      ac_cv_jni_header_path=none
+    fi
+  fi
+])
+
+# get the likely subdirectories for system specific java includes
+case "$host_os" in
+bsdi*)          _JNI_INC_SUBDIRS="bsdos";;
+freebsd*)       _JNI_INC_SUBDIRS="freebsd";;
+darwin*)        _JNI_INC_SUBDIRS="darwin";;
+linux*)         _JNI_INC_SUBDIRS="linux genunix";;
+osf*)           _JNI_INC_SUBDIRS="alpha";;
+solaris*)       _JNI_INC_SUBDIRS="solaris";;
+mingw*)		_JNI_INC_SUBDIRS="win32";;
+cygwin*)	_JNI_INC_SUBDIRS="win32";;
+*)              _JNI_INC_SUBDIRS="genunix";;
+esac
+
+if test "x$ac_cv_jni_header_path" != "xnone"; then
+  # add any subdirectories that are present
+  for JINCSUBDIR in $_JNI_INC_SUBDIRS
+  do
+    if test -d "$_JTOPDIR/include/$JINCSUBDIR"; then
+         JNI_INCLUDE_DIRS="$JNI_INCLUDE_DIRS $_JTOPDIR/include/$JINCSUBDIR"
+    fi
+  done
+fi
+])
+
+# _ACJNI_FOLLOW_SYMLINKS <path>
+# Follows symbolic links on <path>,
+# finally setting variable _ACJNI_FOLLOWED
+# ----------------------------------------
+AC_DEFUN([_ACJNI_FOLLOW_SYMLINKS],[
+# find the include directory relative to the javac executable
+_cur="$1"
+while ls -ld "$_cur" 2>/dev/null | grep " -> " >/dev/null; do
+        AC_MSG_CHECKING([symlink for $_cur])
+        _slink=`ls -ld "$_cur" | sed 's/.* -> //'`
+        case "$_slink" in
+        /*) _cur="$_slink";;
+        # 'X' avoids triggering unwanted echo options.
+        *) _cur=`echo "X$_cur" | sed -e 's/^X//' -e 's:[[^/]]*$::'`"$_slink";;
+        esac
+        AC_MSG_RESULT([$_cur])
+done
+_ACJNI_FOLLOWED="$_cur"
+])# _ACJNI
diff --git a/src/app/main/main.c b/src/app/main/main.c
index a8fc3a93807..6c8ae02e57f 100644
--- a/src/app/main/main.c
+++ b/src/app/main/main.c
@@ -34,6 +34,9 @@
 #include "core/or/connection_or.h"
 #include "core/or/relay.h"
 #include "core/or/status.h"
+#ifdef HAVE_JNI_H
+#include "feature/api/org_torproject_jni_TorService.h"
+#endif
 #include "feature/api/tor_api.h"
 #include "feature/api/tor_api_internal.h"
 #include "feature/client/addressmap.h"
diff --git a/src/core/mainloop/connection.c b/src/core/mainloop/connection.c
index cd71bb8c3a4..bfc8170b77e 100644
--- a/src/core/mainloop/connection.c
+++ b/src/core/mainloop/connection.c
@@ -138,6 +138,18 @@
 #include <sys/un.h>
 #endif
 
+/**
+ * Override log_err and log_warn here since stdout/stderr don't work
+ * https://gitlab.torproject.org/tpo/core/tor/-/issues/32036
+ */
+#ifdef __ANDROID__
+#include <android/log.h>
+#define log_err(ignored, ...) \
+  __android_log_print(ANDROID_LOG_ERROR, "Tor-connection", ##__VA_ARGS__)
+#define log_warn(ignored, ...) \
+  __android_log_print(ANDROID_LOG_WARN, "Tor-connection", ##__VA_ARGS__)
+#endif // __ANDROID__
+
 #include "feature/dircommon/dir_connection_st.h"
 #include "feature/control/control_connection_st.h"
 #include "core/or/entry_connection_st.h"
diff --git a/src/feature/api/.may_include b/src/feature/api/.may_include
index 424c745c125..4bcc583049e 100644
--- a/src/feature/api/.may_include
+++ b/src/feature/api/.may_include
@@ -1 +1,2 @@
 *.h
+*.c
diff --git a/src/feature/api/include.am b/src/feature/api/include.am
index 8d490458d42..4514026eb19 100644
--- a/src/feature/api/include.am
+++ b/src/feature/api/include.am
@@ -3,9 +3,17 @@
 LIBTOR_APP_A_SOURCES += 				\
 	src/feature/api/tor_api.c
 
+if USE_JNI
+LIBTOR_APP_A_SOURCES += src/feature/api/org_torproject_jni_TorService.c
+endif
+
 # ADD_C_FILE: INSERT HEADERS HERE.
 noinst_HEADERS +=					\
 	src/feature/api/tor_api_internal.h
 
 # This may someday want to be an installed file?
 noinst_HEADERS += src/feature/api/tor_api.h
+
+if USE_JNI
+noinst_HEADERS += src/feature/api/org_torproject_jni_TorService.h
+endif
diff --git a/src/feature/api/org_torproject_jni_TorService.c b/src/feature/api/org_torproject_jni_TorService.c
new file mode 100644
index 00000000000..ff587442581
--- /dev/null
+++ b/src/feature/api/org_torproject_jni_TorService.c
@@ -0,0 +1,328 @@
+/* Copyright (c) 2019, Matthew Finkel.
+ * Copyright (c) 2019, Hans-Christoph Steiner.
+ * Copyright (c) 2007-2019, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+#include "org_torproject_jni_TorService.h"
+#include "orconfig.h"
+#include "lib/malloc/malloc.h"
+#include "app/main/shutdown.h"
+#include "feature/api/tor_api.h"
+#include "feature/api/tor_api_internal.h"
+
+
+#include <jni.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <poll.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#ifdef HAVE_SYS_UN_H
+#include <sys/socket.h>
+#include <sys/un.h>
+#endif
+
+#ifdef __ANDROID__
+#include <android/log.h>
+#define fprintf(ignored, ...)                                           \
+  __android_log_print(ANDROID_LOG_ERROR, "Tor-api", ##__VA_ARGS__)
+#endif // __ANDROID__
+
+/* with JNI, unused parameters are inevitable, suppress the warnings */
+#define UNUSED(x) (void)(x)
+
+static char **argv = NULL;
+static int argc = 0;
+
+static jfieldID
+GetConfigurationFieldID(JNIEnv *env, jclass torApiClass)
+{
+  return (*env)->GetFieldID(env, torApiClass, "torConfiguration", "J");
+}
+
+static jlong
+GetConfigurationObject(JNIEnv *env, jobject thisObj)
+{
+  jclass torApiClass = (*env)->GetObjectClass(env, thisObj);
+  if (torApiClass == NULL) {
+    fprintf(stderr, "GetObjectClass returned NULL\n");
+    return 0;
+  }
+
+  jfieldID torConfigurationField = GetConfigurationFieldID(env, torApiClass);
+  if (torConfigurationField == NULL) {
+    fprintf(stderr, "The fieldID is NULL\n");
+    return 0;
+  }
+
+  return (*env)->GetLongField(env, thisObj, torConfigurationField);
+}
+
+static bool
+SetConfiguration(JNIEnv *env, jobject thisObj,
+                 const tor_main_configuration_t* torConfiguration)
+{
+  jclass torApiClass = (*env)->GetObjectClass(env, thisObj);
+  if (torApiClass == NULL) {
+    return false;
+  }
+
+  jfieldID torConfigurationField = GetConfigurationFieldID(env, torApiClass);
+  if (torConfigurationField == NULL) {
+    return false;
+  }
+
+  jlong cfg = (jlong) torConfiguration;
+
+  (*env)->SetLongField(env, thisObj, torConfigurationField, cfg);
+  return true;
+}
+
+static tor_main_configuration_t*
+GetConfiguration(JNIEnv *env, jobject thisObj)
+{
+  jlong torConfiguration = GetConfigurationObject(env, thisObj);
+  if (torConfiguration == 0) {
+    fprintf(stderr, "The long is 0\n");
+    return NULL;
+  }
+
+  return (tor_main_configuration_t *) torConfiguration;
+}
+
+static jfieldID
+GetControlSocketFieldID(JNIEnv * const env, jclass torApiClass)
+{
+  return (*env)->GetFieldID(env, torApiClass, "torControlFd", "I");
+}
+
+static bool
+SetControlSocket(JNIEnv *env, jobject thisObj, int socket)
+{
+  jclass torApiClass = (*env)->GetObjectClass(env, thisObj);
+  if (torApiClass == NULL) {
+    fprintf(stderr, "SetControlSocket: GetObjectClass returned NULL\n");
+    return false;
+  }
+
+  jfieldID controlFieldId = GetControlSocketFieldID(env, torApiClass);
+
+  (*env)->SetIntField(env, thisObj, controlFieldId, socket);
+  return true;
+}
+
+static bool
+CreateTorConfiguration(JNIEnv *env, jobject thisObj)
+{
+  jlong torConfiguration = GetConfigurationObject(env, thisObj);
+  if (torConfiguration == 0) {
+    return false;
+  }
+
+  tor_main_configuration_t *tor_config = tor_main_configuration_new();
+  if (tor_config == NULL) {
+    fprintf(stderr,
+            "Allocating and creating a new configuration structure failed.\n");
+    return false;
+  }
+
+  if (!SetConfiguration(env, thisObj, tor_config)) {
+    tor_main_configuration_free(tor_config);
+    return false;
+  }
+
+  return true;
+}
+
+static bool
+SetCommandLine(JNIEnv *env, jobject thisObj, jobjectArray arrArgv)
+{
+  tor_main_configuration_t *cfg = GetConfiguration(env, thisObj);
+  if (cfg == NULL) {
+    fprintf(stderr, "SetCommandLine: The Tor configuration is NULL!\n");
+    return -1;
+  }
+
+  jsize arrArgvLen = (*env)->GetArrayLength(env, arrArgv);
+  if (arrArgvLen > (INT_MAX-1)) {
+    fprintf(stderr, "Too many args\n");
+    return false;
+  }
+
+  argc = (int) arrArgvLen;
+  argv = (char**) tor_malloc(argc * sizeof(char*));
+  if (argv == NULL) {
+    return false;
+  }
+
+  for (jsize i=0; i<argc; i++) {
+    jobject objElm = (*env)->GetObjectArrayElement(env, arrArgv, i);
+    jstring argElm = (jstring) objElm;
+    const char *arg = (*env)->GetStringUTFChars(env, argElm, NULL);
+    argv[i] = strdup(arg);
+  }
+
+  if (tor_main_configuration_set_command_line(cfg, argc, argv)) {
+    fprintf(stderr, "Setting the command line config failed\n");
+    return false;
+  }
+  return true;
+}
+
+static int
+SetupControlSocket(JNIEnv *env, jobject thisObj)
+{
+  jclass torApiClass = (*env)->GetObjectClass(env, thisObj);
+  if (torApiClass == NULL) {
+    fprintf(stderr, "SetupControlSocket: GetObjectClass returned NULL\n");
+    return false;
+  }
+
+  tor_main_configuration_t *cfg = GetConfiguration(env, thisObj);
+  if (cfg == NULL) {
+    fprintf(stderr, "SetupControlSocket: The Tor configuration is NULL!\n");
+    return false;
+  }
+
+  tor_control_socket_t tcs = tor_main_configuration_setup_control_socket(cfg);
+  fcntl(tcs, F_SETFL, O_NONBLOCK);
+  SetControlSocket(env, thisObj, tcs);
+  return true;
+}
+
+static int
+RunMain(JNIEnv *env, jobject thisObj)
+{
+  tor_main_configuration_t *cfg = GetConfiguration(env, thisObj);
+  if (cfg == NULL) {
+    fprintf(stderr, "RunMain: The Tor configuration is NULL!\n");
+    return -1;
+  }
+
+  int rv = tor_run_main(cfg);
+  if (rv != 0) {
+    fprintf(stderr, "Tor returned with an error\n");
+  } else {
+    printf("Tor returned successfully\n");
+  }
+  return rv;
+}
+
+JNIEXPORT jboolean JNICALL
+Java_org_torproject_jni_TorService_createTorConfiguration
+(JNIEnv *env, jobject thisObj)
+{
+  return CreateTorConfiguration(env, thisObj);
+}
+
+JNIEXPORT jboolean JNICALL
+Java_org_torproject_jni_TorService_mainConfigurationSetCommandLine
+(JNIEnv *env, jobject thisObj, jobjectArray arrArgv)
+{
+  return SetCommandLine(env, thisObj, arrArgv);
+}
+
+JNIEXPORT jboolean JNICALL
+Java_org_torproject_jni_TorService_mainConfigurationSetupControlSocket
+(JNIEnv *env, jobject thisObj)
+{
+  return SetupControlSocket(env, thisObj);
+}
+
+JNIEXPORT void JNICALL
+Java_org_torproject_jni_TorService_mainConfigurationFree
+(JNIEnv *env, jobject thisObj)
+{
+  tor_main_configuration_t *cfg = GetConfiguration(env, thisObj);
+  if (cfg == NULL) {
+    fprintf(stderr, "ConfigurationFree: The Tor configuration is NULL!\n");
+    return;
+  }
+  unset_owning_controller_socket(cfg);
+  tor_main_configuration_free(cfg);
+}
+
+JNIEXPORT jstring JNICALL
+Java_org_torproject_jni_TorService_apiGetProviderVersion
+(JNIEnv *env, jobject _ignore)
+{
+  UNUSED(_ignore);
+  return (*env)->NewStringUTF(env, tor_api_get_provider_version());
+}
+
+JNIEXPORT jint JNICALL
+Java_org_torproject_jni_TorService_runMain
+(JNIEnv *env, jobject thisObj)
+{
+  return RunMain(env, thisObj);
+}
+
+JNIEXPORT void JNICALL
+Java_org_torproject_jni_TorService_torFreeAll
+(JNIEnv *env, jobject _ignore)
+{
+  UNUSED(env);
+  UNUSED(_ignore);
+  tor_free_all(0);
+}
+
+/**
+ * Android does not support UNIX Domain Sockets, but we can fake it by sending
+ * the file descriptor via a java.io.FileDescriptor instance, which can be
+ * used to open streams.  The field "fd" has been in Java forever.  In Android,
+ * they renamed the field in 2008 to "descriptor", back when they did many
+ * silly things like that.  It hasn't changed since then, e.g. Android 1.0.
+ */
+JNIEXPORT jobject JNICALL
+Java_org_torproject_jni_TorService_prepareFileDescriptor
+(JNIEnv *env, jclass _ignore, jstring arg)
+{
+  UNUSED(_ignore);
+  const char *filename = (*env)->GetStringUTFChars(env, arg, NULL);
+  int fd = socket(AF_UNIX, SOCK_STREAM, 0);
+  struct sockaddr_un addr;
+  memset(&addr, 0, sizeof(addr));
+  addr.sun_family = AF_UNIX;
+  strncpy(addr.sun_path, filename, sizeof(addr.sun_path) - 1);
+  (*env)->ReleaseStringUTFChars(env, arg, filename);
+
+  jclass io_exception = (*env)->FindClass(env, "java/io/IOException");
+  if (io_exception == NULL) {
+    return NULL;
+  }
+  if (fd < 0 || connect(fd, (struct sockaddr *) &addr, sizeof(addr)) == -1) {
+    char buf[1024];
+    snprintf(buf, 1023, "%s open: %s", filename, strerror(errno));
+    (*env)->ThrowNew(env, io_exception, buf);
+    return NULL;
+  }
+
+  jclass file_descriptor = (*env)->FindClass(env, "java/io/FileDescriptor");
+  if (file_descriptor == NULL) {
+    return NULL;
+  }
+  jmethodID file_descriptor_init = \
+    (*env)->GetMethodID(env, file_descriptor, "<init>", "()V");
+  if (file_descriptor_init == NULL) {
+    return NULL;
+  }
+  jobject ret = (*env)->NewObject(env, file_descriptor, file_descriptor_init);
+
+#ifdef __ANDROID__
+  jfieldID field_fd = \
+    (*env)->GetFieldID(env, file_descriptor, "descriptor", "I");
+#else  /* !defined(__ANDROID__) */
+  jfieldID field_fd = (*env)->GetFieldID(env, file_descriptor, "fd", "I");
+#endif  /* defined(__ANDROID__) */
+  if (field_fd == NULL) {
+    return NULL;
+  }
+  (*env)->SetIntField(env, ret, field_fd, fd);
+
+  return ret;
+}
diff --git a/src/feature/api/org_torproject_jni_TorService.h b/src/feature/api/org_torproject_jni_TorService.h
new file mode 100644
index 00000000000..4ea0e5f9e20
--- /dev/null
+++ b/src/feature/api/org_torproject_jni_TorService.h
@@ -0,0 +1,82 @@
+/* Copyright (c) 2019, Matthew Finkel.
+ * Copyright (c) 2019, Hans-Christoph Steiner.
+ * Copyright (c) 2007-2019, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+#ifndef ORG_TORPROJECT_JNI_TORSERVICE_H
+#define ORG_TORPROJECT_JNI_TORSERVICE_H
+
+#include <jni.h>
+
+/*
+ * Class:     org_torproject_jni_TorService
+ * Method:    createTorConfiguration
+ * Signature: ()Z
+ */
+JNIEXPORT jboolean JNICALL
+Java_org_torproject_jni_TorService_createTorConfiguration
+(JNIEnv *, jobject);
+
+/*
+ * Class:     org_torproject_jni_TorService
+ * Method:    mainConfigurationSetCommandLine
+ * Signature: ([Ljava/lang/String;)Z
+ */
+JNIEXPORT jboolean JNICALL
+Java_org_torproject_jni_TorService_mainConfigurationSetCommandLine
+(JNIEnv *, jobject, jobjectArray);
+
+/*
+ * Class:     org_torproject_jni_TorService
+ * Method:    mainConfigurationSetupControlSocket
+ * Signature: ()Z
+ */
+JNIEXPORT jboolean JNICALL
+Java_org_torproject_jni_TorService_mainConfigurationSetupControlSocket
+(JNIEnv *, jobject);
+
+/*
+ * Class:     org_torproject_jni_TorService
+ * Method:    mainConfigurationFree
+ * Signature: ()V
+ */
+JNIEXPORT void JNICALL
+Java_org_torproject_jni_TorService_mainConfigurationFree
+(JNIEnv *, jobject);
+
+/*
+ * Class:     org_torproject_jni_TorService
+ * Method:    apiGetProviderVersion
+ * Signature: ()Ljava/lang/String;
+ */
+JNIEXPORT jstring JNICALL
+Java_org_torproject_jni_TorService_apiGetProviderVersion
+(JNIEnv *, jobject);
+
+/*
+ * Class:     org_torproject_jni_TorService
+ * Method:    runMain
+ * Signature: ()I
+ */
+JNIEXPORT jint JNICALL
+Java_org_torproject_jni_TorService_runMain
+(JNIEnv *, jobject);
+
+/*
+ * Class:     org_torproject_jni_TorService
+ * Method:    torFreeAll
+ * Signature: ()V
+ */
+JNIEXPORT void JNICALL
+Java_org_torproject_jni_TorService_torFreeAll
+(JNIEnv *, jobject);
+
+/*
+ * Class:     org_torproject_jni_TorService
+ * Method:    prepareFileDescriptor
+ */
+JNIEXPORT jobject JNICALL
+Java_org_torproject_jni_TorService_prepareFileDescriptor
+(JNIEnv *env, jclass, jstring);
+
+#endif /* !defined(ORG_TORPROJECT_JNI_TORSERVICE_H) */
diff --git a/src/feature/api/tor_api.c b/src/feature/api/tor_api.c
index 88e91ebfd5e..6396b62b51a 100644
--- a/src/feature/api/tor_api.c
+++ b/src/feature/api/tor_api.c
@@ -100,6 +100,17 @@ tor_main_configuration_set_command_line(tor_main_configuration_t *cfg,
   return 0;
 }
 
+/*
+ * Temporary workaround for a double close crash with fdsan.
+ * See: https://github.com/guardianproject/tor-android/issues/57
+ * Should be removed once fixed in Tor upstream.
+ */
+void
+unset_owning_controller_socket(tor_main_configuration_t* t)
+{
+    t->owning_controller_socket = -1;
+}
+
 tor_control_socket_t
 tor_main_configuration_setup_control_socket(tor_main_configuration_t *cfg)
 {
@@ -167,3 +178,7 @@ tor_main(int argc, char *argv[])
   tor_main_configuration_free(cfg);
   return rv;
 }
+
+#ifdef HAVE_JNI_H
+#include "org_torproject_jni_TorService.c"
+#endif
diff --git a/src/feature/api/tor_api_internal.h b/src/feature/api/tor_api_internal.h
index 5075922676f..719685cb137 100644
--- a/src/feature/api/tor_api_internal.h
+++ b/src/feature/api/tor_api_internal.h
@@ -31,4 +31,11 @@ struct tor_main_configuration_t {
   tor_socket_t owning_controller_socket;
 };
 
+/*
+ * Temporary workaround for a double close crash with fdsan.
+ * See: https://github.com/guardianproject/tor-android/issues/57
+ * Should be removed once fixed in Tor upstream.
+ */
+void unset_owning_controller_socket(tor_main_configuration_t* t);
+
 #endif /* !defined(TOR_API_INTERNAL_H) */
